text/0000-repr-type-aliases.md

```diff
@@ -0,0 +1,84 @@
+- Feature Name: `repr_type_aliases`
+- Start Date: 2024-06-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Primitive representations on enums now accept type aliases, meaning that in addition to primitives like `#[repr(u32)]`, `#[repr(type = core::ffi::c_int)]` and `#[repr(type = my_type)]` are now accepted.
+
+# Motivation
+[motivation]: #motivation
+
+For the same reasons why type aliases are useful, having type aliases in `repr` attributes would also be useful. A few examples:
+
+* Types depend on an external API whose exact size may be uncertain. (e.g. `core::ffi::c_int`, `gl::types::GLsizei`)
+* An internal API might want to be able to easily change a type later.
+* The intent behind a type alias may be clearer than simply using the primitive directly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enums allow `#[repr(type = ...)]` attributes to offer an explicit discriminant type. (`...` can be any primitive integer type, like `u8`, `i32`, or `usize`, but not `char`.) If all variants of the enum are unit variants, this means that the enum will be easily castable to `type` using `as`. Otherwise, the discriminant will still be of the specified type, but unsafe code is required to actually access it.
+
+To ensure compatibility, the `#[repr(type = ...)]` form is required if the type is not one of the known primitive types. Note that this form is not necessarily equivalent to using the primitive representations directly, since shadowing is possible; for example, if you did `type u32 = u8` and then `#[repr(type = u32)]`, this would be equivalent to `#[repr(u8)]`, not `#[repr(u32)]`.
+
+You can use any type alias in the `repr` attribute, but it *must* be an alias to an accepted primitive type like `u8` or `i32`, and cannot be a pointer, reference, struct, etc.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `repr` attribute now accepts a `type = ...` argument to indicate a resolved path instead of a well-known primitive type. If those the path resolves to a type alias to a valid primitive type which can be used in the `repr` attribute, that will be used as the actual discriminant representation.
+
+An additional, automatically-applicable lint should be added that warns a user if they use `type = ...` for a well-known primitive type, since adding `type = ` instead of using the type directly introduces the possibility of shadowing. (For example, `#[repr(type = u32)]` becomes `#[repr(u32)]`.)
```

Comment:
I'm a little skeptical that such a lint should be added and especially skeptical of its automatically-applicable suggested fix.

First, I am skeptical that such a lint should be added because macro authors will almost certainly want to standardize on the general `type = Ty` form. Such a lint, emitted after macro expansion, will mean that the internal details of macro expansion (i.e., which `repr` form the macro emits) will leak to its use site.

If `u32` is shadowed to mean something *other* than `::core::primitive::u32`, the problems that might arise from that are far-reaching — hardly limited to `repr(type = Ty)`. The actionable location for such a lint is wherever that shadowing occurs. 

Second, I am skeptical of the suggested fix here. As a matter of style, I think I will probably prefer to consistently use the `type = Ty` syntax, even in cases where it might be abbreviated. I don't think Rust typically uses lints to enforce matters of style.

Reply:
So, I've updated the suggestion here, and hoping that it makes more sense: instead of an automatically-applicable lint, the suggestion is now to add an allow-by-default or `clippy::pedantic` lint which enforces absolute paths for the type alias, for people who are concerned about shadowing.

This is in addition to the automtically-applicable suggestion where you do `#[repr(MyType)]` to instead suggest `#[repr(discriminant = MyType)]`, where `MyType` isn't a valid argument to `repr`.